### PR TITLE
Multivariate discrete distributions: hypergeometric, Polya, plus a shared spectral CDF core

### PR DESCRIPTION
**Pull request by regislebrun**
# Multivariate discrete distributions: hypergeometric, Polya, plus a shared spectral CDF core

## Summary

This branch reworks the multivariate discrete distributions around a single
truncated-generating-function spectral machinery, and adds two new experimental
distributions that reuse it:

- a shared private core (`PoissonFormula.hxx`) implementing the DFT-based
  cumulative/probability computations for multivariate discrete counts, with
  log1p/expm1 based kernels;
- the new experimental **MultivariateHypergeometric** distribution and factory;
- the new experimental **MultivariatePolya** distribution and factory;
- bug fixes to the existing **Multinomial** class (below-mode truncation of the
  generating function, negative interval bounds, CDF/conditional edge cases).

The spectral (generating-function) engine is kept for all three distributions:
a head-to-head benchmark against a from-scratch Poisson-summation reference
showed it stays accurate (CDF to ~1e-12 relative at n=600, well below spatial
underflow) while being faster and needing no adaptive fallback. Detailed
accuracy/timing figures and the benchmark rationale are documented in
`doc/multivariate_discrete/` on this branch (working notes, not committed).

## Changes

- **Distrib|MultivariateDiscrete: share the Poisson-formula spectral core**
  Adds the private `PoissonFormula.hxx` helper (not installed), wires the two
  new experimental classes into the build, ResourceMap, ChangeLog and user manual.

- **Multinomial|Distrib: fix CDF, probability and conditional edge cases**
  Delegates Multinomial to the shared spectral core; fixes the below-mode cap of
  `computeLocalPhi`, negative interval bounds and CDF/conditional edge cases.
  Python tests moved to the assert-based paradigm.

- **MultivariateHypergeometric|Distrib: add the multivariate hypergeometric distribution and factory**
  New experimental distribution with factory, C++ and Python tests, docs.

- **MultivariatePolya|Distrib: add the multivariate Polya distribution and factory**
  New experimental distribution with factory, C++ and Python tests, docs.

## Notes

- New classes land under `openturns.experimental`.
- Tests added in the Python folder (assert-based). Existing C++ tests are kept
  and keep passing.
- Accuracy is validated against an independent from-scratch Poisson-summation
  reference (mpmath) up to 4D, defective support, n up to 600:
  PMF to ~1e-16, CDF to ~8e-15 (defective 2D) / ~9e-12 (n=600).

## Summary by CodeRabbit

- **New Features**
  - Added experimental multivariate hypergeometric and multivariate Polya distributions.
  - Added factories for constructing and estimating both distributions from samples.
  - Exposed the distributions and factories through the experimental Python module.
  - Added sampling, probability, marginal, support, moment, and quantile operations.

- **Bug Fixes**
  - Improved edge-case handling and numerical reliability for multinomial calculations.

- **Documentation**
  - Added user documentation, examples, parameter guidance, experimental-status notices, and a bibliography reference.

### CODE REVIEW

File: lib/src/Uncertainty/Distribution/Multinomial.cxx [hunk 1]
```diff
@@ -467,6 +397,12 @@ Scalar Multinomial::computeProbability(const Interval & interval) const
     upper.add(n_);
     return Multinomial(n_, p).computeProbability(Interval(lower, upper));
   }
+  // The box is empty if one of its upper bounds is strictly negative or one of
+  // its lower bounds exceeds the support upper bound n_
+  for (UnsignedInteger i = 0; i < dimension_; ++i)
+  {
+    if ((upper[i] < -supportEpsilon_) || (lower[i] > static_cast<Scalar>(n_) + supportEpsilon_)) return 0.0;
```
[the remark this answers is not in the document]
Comment (regislebrun) [reply lib/src/Uncertainty/Distribution/Multinomial.cxx:405]: fixed

File: lib/src/Uncertainty/Distribution/Multinomial.cxx [hunk 2]
```diff
@@ -552,20 +469,20 @@ Scalar Multinomial::computeConditionalPDF(const Scalar x,
...
+    const Scalar yRounded = round(yI);
+    if ((yRounded < 0.0) || (std::abs(yI - yRounded) > supportEpsilon_)) throw InvalidArgumentException(HERE) << "Error: the conditioning vector has non-integer values";
     sumY += yI;
-    intY[i] = intYI;
     sumP += p_[i];
   }
   if (sumY > n_) throw InvalidArgumentException(HERE) << "Error: the conditioning vector has a sum of components greater than the allowed range.";
   if (sumP >= 1.0 - supportEpsilon_) return (x == 0.0 ? 1.0 : 0.0);
+  const Scalar xRounded = round(x);
+  if ((xRounded < 0.0) || (std::abs(x - xRounded) > supportEpsilon_)) return 0.0;
```
[the remark this answers is not in the document]
Comment (regislebrun) [reply lib/src/Uncertainty/Distribution/Multinomial.cxx:486]: fixed

File: lib/src/Uncertainty/Distribution/MultivariateHypergeometric.cxx [hunk 3]
```diff
@@ -0,0 +1,825 @@
...
+void MultivariateHypergeometric::setParameter(const Point & parameter)
+{
+  const UnsignedInteger dimension = getDimension();
+  if (parameter.getDimension() != (dimension + 1))
+    throw InvalidArgumentException(HERE) << "MultivariateHypergeometric expected " << (dimension + 1) << " parameters, got " << parameter.getDimension();
+  if (parameter[0] != std::round(parameter[0])) throw InvalidArgumentException(HERE) << "the MultivariateHypergeometric first parameter n must be an integer, got " << parameter[0];
+  setN(parameter[0]);
+  Point colors(dimension);
+  std::copy(parameter.begin() + 1, parameter.end(), colors.begin());
+  setColors(colors);
```
[the remark this answers is not in the document]
Comment (regislebrun) [reply lib/src/Uncertainty/Distribution/MultivariateHypergeometric.cxx]: fixed

File: lib/src/Uncertainty/Distribution/MultivariateHypergeometric.cxx [hunk 4]
```diff
@@ -0,0 +1,825 @@
...
+Scalar MultivariateHypergeometric::getSmallA() const
+{
+  return smallA_;
+}
+
+/* Eta accessor */
+void MultivariateHypergeometric::setEta(const Scalar eta)
+{
+  eta_ = eta;
+}
```
[the remark this answers is not in the document]
Comment (regislebrun) [reply lib/src/Uncertainty/Distribution/MultivariateHypergeometric.cxx:799]: fixed

File: lib/src/Uncertainty/Distribution/MultivariateHypergeometricFactory.cxx [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+    if (K < n) K = n;
+    UnsignedInteger sumColors = 0;
+    for (UnsignedInteger j = 0; j < dimension; ++j)
+    {
+      colors[j] = (proportions[j] == 0.0 ? 0.0 : std::max(1.0, std::round(K * proportions[j])));
+      sumColors += static_cast<UnsignedInteger>(colors[j]);
+    }
+    feasible = (sumColors >= n);
+    ++K;
+  }
```
[the remark this answers is not in the document]
Comment (regislebrun) [reply lib/src/Uncertainty/Distribution/MultivariateHypergeometricFactory.cxx:118]: fixed

File: lib/src/Uncertainty/Distribution/MultivariatePolya.cxx [hunk 6]
```diff
@@ -0,0 +1,873 @@
...
+  return description;
+}
+
+void MultivariatePolya::setParameter(const Point & parameter)
+{
+  const UnsignedInteger dimension = getDimension();
+  if (parameter.getDimension() != (dimension + 2))
+    throw InvalidArgumentException(HERE) << "MultivariatePolya expected " << (dimension + 2) << " parameters, got " << parameter.getDimension();
+  if (parameter[0] != std::round(parameter[0])) throw InvalidArgumentException(HERE) << "the MultivariatePolya first parameter n must be an integer, got " << parameter[0];
+  setN(parameter[0]);
```
[the remark this answers is not in the document]
Comment (regislebrun) [reply lib/src/Uncertainty/Distribution/MultivariatePolya.cxx:743]: fixed

File: lib/src/Uncertainty/Distribution/MultivariatePolya.cxx [hunk 7]
```diff
@@ -0,0 +1,873 @@
...
+  return eta_;
+}
+
+/* Compute the scalar quantile of the 1D multivariate Polya distribution */
+Scalar MultivariatePolya::computeScalarQuantile(const Scalar prob,
+                                                const Bool tail) const
+{
+  // The single component of a 1D multivariate Polya distribution is equal to n_
+  return 1.0 * n_;
+}
```
[the remark this answers is not in the document]
Comment (regislebrun) [reply lib/src/Uncertainty/Distribution/MultivariatePolya.cxx:863]: fixed

File: lib/src/Uncertainty/Distribution/PoissonFormula.hxx [hunk 8]
```diff
@@ -0,0 +1,403 @@
...
+    // The whole window lies above the mode: the terms p(k) z^(k - a) increase
+    // as k decreases toward the mode, so start from b.
+    Complex value(kernel.pmfAt(b) * std::pow(z, 1.0 * (b - a)));
+    Complex term(value);
+    for (UnsignedInteger i = b; i > a; --i)
+    {
+      term = kernel.down(term, i);
+      value += term;
+    }
+    return value;
```
[the remark this answers is not in the document]
Comment (regislebrun) [reply lib/src/Uncertainty/Distribution/PoissonFormula.hxx:325]: fixed

File: python/src/MultivariatePolyaFactory_doc.i [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+R"RAW(Multivariate Polya factory.
+
+.. warning::
+    This class is experimental and likely to be modified in future releases.
+    To use it, import the ``openturns.experimental`` submodule.
+
+Notes
+-----
+The parameters :math:`N`, :math:`\\vect{\\alpha}` and :math:`c` are estimated
+from a sample :math:`(\\vect{x}^1, \\ldots, \\vect{x}^m)` using moment
```
[the remark this answers is not in the document]
Comment (regislebrun) [reply python/src/MultivariatePolyaFactory_doc.i]: fixed

File: python/test/t_MultivariateHypergeometricFactory_std.py [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+
+# Error cases
+
+
+def check_bad(arg, msg):
+    try:
+        factory.buildAsMultivariateHypergeometric(arg)
+        raise AssertionError("should have thrown for " + msg)
+    except Exception:
+        pass
```
[the remark this answers is not in the document]
Comment (regislebrun) [reply python/test/t_MultivariateHypergeometricFactory_std.py]: fixed

File: python/src/MultivariatePolya_doc.i [hunk 11]
```diff
@@ -0,0 +1,230 @@
...
+        \Expect{X_i} & = & N \\frac{\\alpha_i}{q}, \quad i = 1, \ldots, d \\
+        \Var{X_i} & = & N \\frac{\\alpha_i}{q} \\left(1 - \\frac{\\alpha_i}{q}\\right) \\frac{N + \\theta}{1 + \\theta}, \quad \\theta = \\frac{q}{c}, \quad i = 1, \ldots, d \\
+        \Cov{X_i, X_j} & = & - N \\frac{\\alpha_i \\alpha_j}{q^2} \\frac{N + \\theta}{1 + \\theta}, \quad i, j = 1, \ldots, d, i \\neq j
+    \end{eqnarray*}
+
+where the overdispersion factor :math:`(N + \\theta) / (1 + \\theta)` is equal
+to 1 in the degenerate multinomial case :math:`c = 0`.
+
+The marginal distribution of a category is a beta-binomial
+:math:`\mathrm{X_i} \\sim \mathrm{BetaBinomial}(N, \\alpha_i, q - \\alpha_i)`,
```
[the remark this answers is not in the document]
Comment (regislebrun) [reply python/src/MultivariatePolya_doc.i]: fixed

File: lib/src/Uncertainty/Distribution/MultivariateHypergeometric.cxx [hunk 12]
```diff
@@ -0,0 +1,837 @@
...
+    const Scalar c = colors_[0];
+    // Deterministic case
+    if (total_ == n_) return (k >= c - supportEpsilon_ ? 1.0 : 0.0);
+    if (c == total_) return (k >= n_ - supportEpsilon_ ? 1.0 : 0.0);
+    if (c == 0.0) return (k >= 0.0 ? 1.0 : 0.0);
+    // Support of the marginal: [max(0, c + n_ - total_), min(c, n_)]
+    if (k < c + n_ - total_ - supportEpsilon_) return 0.0;
+    if (k >= std::min(c, 1.0 * n_) + supportEpsilon_) return 1.0;
+    if (k < -supportEpsilon_) return 0.0;
+    return DistFunc::pHypergeometric(static_cast<UnsignedInteger>(total_), static_cast<UnsignedInteger>(c), n_, static_cast<UnsignedInteger>(floor(k)), false);
```
[the remark this answers is not in the document]
Comment (regislebrun) [reply lib/src/Uncertainty/Distribution/MultivariateHypergeometric.cxx]: fixed

File: lib/src/Uncertainty/Distribution/MultivariateHypergeometric.cxx [hunk 13]
```diff
@@ -0,0 +1,837 @@
...
+    {
+      const UnsignedInteger xI = DistFunc::rHypergeometric(static_cast<UnsignedInteger>(remainingTotal), static_cast<UnsignedInteger>(colors_[i]), remainingN);
+      sample(k, i) = xI;
+      remainingN -= xI;
+      if (remainingN == 0) break;
+      remainingTotal -= colors_[i];
+    }
+  }
+  return sample;
+}
```
[the remark this answers is not in the document]
Comment (regislebrun) [reply lib/src/Uncertainty/Distribution/MultivariateHypergeometric.cxx:206]: fixed

File: lib/src/Uncertainty/Distribution/MultivariateHypergeometric.cxx [hunk 14]
```diff
@@ -0,0 +1,837 @@
...
+  for (UnsignedInteger i = 0; i < dimension; ++i)
+  {
+    // If the given point does not cover any point of the support, return 0.0
+    if (point[i] < 0.0) return 0.0;
+    allZero = allZero && (point[i] <= supportEpsilon_);
+  }
+  // Now we know all components are non-negative, so conversion is safe
+  const Indices kPoint(point.begin(), point.end());
+  for (UnsignedInteger i = 0; i < dimension; ++i)
+    sumX += kPoint[i];
```
[the remark this answers is not in the document]
Comment (regislebrun) [reply lib/src/Uncertainty/Distribution/MultivariateHypergeometric.cxx:334]: fixed

File: lib/src/Uncertainty/Distribution/MultivariateHypergeometric.cxx [hunk 15]
```diff
@@ -0,0 +1,837 @@
...
+
+/* Get the distribution of the marginal distribution corresponding to indices dimensions */
+Distribution MultivariateHypergeometric::getMarginal(const Indices & indices) const
+{
+  const UnsignedInteger dimension = getDimension();
+  if (!indices.check(dimension)) throw InvalidArgumentException(HERE) << "The indices of a marginal distribution must be in the range [0, dim-1] and must be different";
+  // Special case for dimension 1
+  if (dimension == 1) return clone();
+  // Special case for a full marginal
+  if (indices.getSize() == dimension) return clone();
```
[the remark this answers is not in the document]
Comment (regislebrun) [reply lib/src/Uncertainty/Distribution/MultivariateHypergeometric.cxx]: fixed

File: lib/src/Uncertainty/Distribution/MultivariateHypergeometric.cxx [hunk 16]
```diff
@@ -0,0 +1,837 @@
...
+  // the shrink case, eg (5, [10]) -> (4, [5]), where a componentwise reordering fails
+  Point colors(dimension);
+  std::copy(parameter.begin() + 1, parameter.end(), colors.begin());
+  const Scalar eta = eta_;
+  const Scalar smallA = smallA_;
+  const Scalar w = getWeight();
+  const Description d(getDescription());
+  *this = MultivariateHypergeometric(static_cast<UnsignedInteger>(parameter[0]), colors);
+  eta_ = eta;
+  smallA_ = smallA;
```
[the remark this answers is not in the document]
Comment (regislebrun) [reply lib/src/Uncertainty/Distribution/MultivariateHypergeometric.cxx]: fixed

File: lib/src/Uncertainty/Distribution/MultivariatePolya.cxx [hunk 17]
```diff
@@ -0,0 +1,884 @@
...
+
+/* Get the distribution of the marginal distribution corresponding to indices dimensions */
+Distribution MultivariatePolya::getMarginal(const Indices & indices) const
+{
+  const UnsignedInteger dimension = getDimension();
+  if (!indices.check(dimension)) throw InvalidArgumentException(HERE) << "The indices of a marginal distribution must be in the range [0, dim-1] and must be different";
+  // Special case for dimension 1
+  if (dimension == 1) return clone();
+  // Special case for a full marginal
+  if (indices.getSize() == dimension) return clone();
```
Comment (jschueller) [inline lib/src/Uncertainty/Distribution/MultivariatePolya.cxx]: alpha may be reordered if indices is not the identity ?

Comment (regislebrun) [reply lib/src/Uncertainty/Distribution/MultivariatePolya.cxx]: fixed

File: python/src/Multinomial_doc.i [hunk 18]
```diff
@@ -165,9 +180,12 @@ smallA : float
 
 Parameters
 ----------
-smallA : float
+smallA : float, :math:`smallA \\geq 0`
```
Comment (jschueller) [inline python/src/Multinomial_doc.i]: single backslash, use RAW markers

File: python/test/t_MultivariatePolyaFactory_std.py [hunk 19]
```diff
@@ -0,0 +1,58 @@
...
+ott.assert_almost_equal(degenerate.getC(), 0.0)
+
+
+def check_bad(arg, msg):
+    try:
+        factory.buildAsMultivariatePolya(arg)
+    except TypeError:
+        pass
+    else:
+        raise AssertionError("should have thrown for " + msg)
```
Comment (jschueller) [inline python/test/t_MultivariatePolyaFactory_std.py]: use assert_raises

File: python/test/t_MultivariateHypergeometricFactory_std.py [hunk 20]
```diff
@@ -0,0 +1,54 @@
...
+
+# Error cases
+
+
+def check_bad(arg, msg):
+    try:
+        factory.buildAsMultivariateHypergeometric(arg)
+    except TypeError:
+        pass
+    else:
```
Comment (jschueller) [inline python/test/t_MultivariateHypergeometricFactory_std.py]: assert_raises

File: python/test/t_MultivariatePolya_std.py [hunk 21]
```diff
@@ -0,0 +1,230 @@
...
+ott.assert_almost_equal(
+    dist_reg.computeProbability(
+        ot.Interval([21.0, 0.0, 0.0], [60.0, 300.0, 300.0])
+    ),
+    0.09581,
+    1e-4,
+    1e-4,
+)
+
+print("OK")
```
Comment (jschueller) [inline python/test/t_MultivariatePolya_std.py:238]: can we use DistributionValidation ?

File: python/test/t_MultivariateHypergeometric_std.py [hunk 22]
```diff
@@ -0,0 +1,262 @@
...
+ott.assert_almost_equal(dist_eta.getEta(), 0.5)
+dist_eta_ref = otexp.MultivariateHypergeometric(4, [3, 3, 3])
+dist_eta_ref.setEta(0.5)
+ott.assert_almost_equal(dist_eta.computeCDF([2, 2, 2]), dist_eta_ref.computeCDF([2, 2, 2]), 1e-12)
+
+# computeLogPDF on-support point matches log(computePDF)
+for a, _ in mass3:
+    ott.assert_almost_equal(dist3.computeLogPDF(a), math.log(dist3.computePDF(a)), 1e-12, 1e-15)
+
+print("OK")
```
Comment (jschueller) [inline python/test/t_MultivariateHypergeometric_std.py:270]: can we use DistributionValidation ?

File: lib/src/Uncertainty/Distribution/MultivariatePolya.cxx [hunk 23]
```diff
@@ -0,0 +1,894 @@
...
+      }
+      for (UnsignedInteger i = 0; i < dimension; ++i) pTilde[i] /= gammaSum;
+    }
+    else
+    {
+      pTilde = p;
+    }
+    const Point realization(drawMultinomial(n_, pTilde));
+    for (UnsignedInteger i = 0; i < dimension; ++i) sample(k, i) = realization[i];
+  }
```
Comment (jschueller) [inline lib/src/Uncertainty/Distribution/MultivariatePolya.cxx:256]: like multivariatehypergeometric:
sample.setName(getName());
  sample.setDescription(getDescription());

File: lib/src/Uncertainty/Distribution/MultivariatePolya.cxx [hunk 24]
```diff
@@ -0,0 +1,894 @@
...
+  Bool allZero = true;
+  UnsignedInteger sumX = 0;
+  for (UnsignedInteger i = 0; i < dimension; ++i)
+  {
+    // If the given point does not cover any point of the support, return 0.0
+    if (point[i] < 0.0) return 0.0;
+    allZero = allZero && (point[i] <= supportEpsilon_);
+  }
+  // Now we know all components are non-negative, so conversion is safe
+  const Indices kPoint(point.begin(), point.end());
```
Comment (jschueller) [inline lib/src/Uncertainty/Distribution/MultivariatePolya.cxx]: no need for clamping like done in MultivariateHypergeometric ?

File: lib/src/Uncertainty/Distribution/MultivariatePolya.cxx [hunk 25]
```diff
@@ -0,0 +1,894 @@
...
+/* SmallA accessor */
+Scalar MultivariatePolya::getSmallA() const
+{
+  return smallA_;
+}
+
+/* Eta accessor */
+void MultivariatePolya::setEta(const Scalar eta)
+{
+  eta_ = eta;
```
Comment (jschueller) [inline lib/src/Uncertainty/Distribution/MultivariatePolya.cxx:845]: check eta>0 ?

File: lib/src/Uncertainty/Distribution/MultivariateHypergeometric.cxx [hunk 26]
```diff
@@ -0,0 +1,857 @@
...
+      const Scalar logFactor = SpecFunc::LogGamma(1.0 * colorJ + 1.0) - SpecFunc::LogGamma(1.0 * x + 1.0) - SpecFunc::LogGamma(1.0 * (colorJ - x) + 1.0);
+      atom[depth] = 1.0 * x;
+      norm += x;
+      logProd += logFactor;
+      visit(depth + 1);
+      logProd -= logFactor;
+      norm -= x;
+    }
+  };
+  visit(0);
```
Comment (jschueller) [inline lib/src/Uncertainty/Distribution/MultivariateHypergeometric.cxx:526]: so this grows exponentially right ?

File: lib/src/Uncertainty/Distribution/MultivariateHypergeometric.cxx [hunk 27]
```diff
@@ -0,0 +1,857 @@
...
+  const UnsignedInteger dimension = getDimension();
+  const UnsignedInteger colorsSize = colors_.getDimension();
+  // The number of draws may be set before the population counts, in which case
+  // the range is left empty until setColors fills the colors vector
+  Point lowerBound(dimension);
+  Point upperBound(dimension);
+  const UnsignedInteger breakpoint = std::min(dimension, colorsSize);
+  for (UnsignedInteger i = 0; i < breakpoint; ++i)
+  {
+    lowerBound[i] = 0.0;
```
Comment (jschueller) [inline lib/src/Uncertainty/Distribution/MultivariateHypergeometric.cxx]: seems there is a better lower bound formula: max(0,n-(total-c_i))

File: lib/src/Uncertainty/Distribution/PoissonFormula.hxx [hunk 28]
```diff
@@ -0,0 +1,424 @@
...
+    term = kernel.up(term, i);
+    value -= term;
+  }
+  return value;
+} // localPhi
+
+/* Telescoped root of unity filter extracting P(W = n) from the generating
+   function, used by the CDF computation. The sum is alternating and stops as
+   soon as the partial sums become stable to a relative precision. */
+inline Scalar computeCDF(const std::function<Complex(const Complex &)> & phi,
```
Comment (jschueller) [inline lib/src/Uncertainty/Distribution/PoissonFormula.hxx]: this mayb be faster using templates:
```
template <typename Function>
inline Scalar computeCDF(const Function & phi, ...
```

File: lib/src/Uncertainty/Distribution/PoissonFormula.hxx [hunk 29]
```diff
@@ -0,0 +1,424 @@
...
+    if (dv < SpecFunc::Precision * dv0) break;
+    sign = -sign;
+  }
+  // Due to round-off errors, the computed CDF can be slightly below 0 or over 1.
+  return SpecFunc::Clip01(value * normalization);
+} // computeCDF
+
+/* Root of unity filter extracting the probability content of a box from the
+   generating function, with the diametral term handling the odd coefficient. */
+inline Scalar computeProbability(const std::function<Complex(const Complex &)> & phi,
```
Comment (jschueller) [inline lib/src/Uncertainty/Distribution/PoissonFormula.hxx]: here too

File: lib/src/Uncertainty/Distribution/Multinomial.cxx
Comment (jschueller) [inline lib/src/Uncertainty/Distribution/Multinomial.cxx:244]: there might be degenerate cases to consider when p_[i] = 0

**Review by jschueller — APPROVED** [review]
impressive, but difficult to review :]

I suggest to do a pass with a strong LLM model

**Comment by regislebrun** [comment]
@jschueller There is nothing more than what is exposed here: https://link.springer.com/article/10.1007/s11222-012-9334-8

You can read an extended version in https://theses.hal.science/tel-00913510, chapter 7

File: python/src/Multinomial_doc.i (hunk 18 above)
Comment (regislebrun) [reply python/src/Multinomial_doc.i]: fixed

File: python/test/t_MultivariatePolyaFactory_std.py (hunk 19 above)
Comment (regislebrun) [reply python/test/t_MultivariatePolyaFactory_std.py]: fixed

File: python/test/t_MultivariateHypergeometricFactory_std.py (hunk 20 above)
Comment (regislebrun) [reply python/test/t_MultivariateHypergeometricFactory_std.py]: fixed

File: python/test/t_MultivariatePolya_std.py (hunk 21 above)
Comment (regislebrun) [reply python/test/t_MultivariatePolya_std.py:238]: fixed

File: python/test/t_MultivariateHypergeometric_std.py (hunk 22 above)
Comment (regislebrun) [reply python/test/t_MultivariateHypergeometric_std.py:270]: fixed

File: lib/src/Uncertainty/Distribution/MultivariatePolya.cxx (hunk 23 above)
Comment (regislebrun) [reply lib/src/Uncertainty/Distribution/MultivariatePolya.cxx:256]: fixed

File: lib/src/Uncertainty/Distribution/MultivariatePolya.cxx (hunk 24 above)
Comment (regislebrun) [reply lib/src/Uncertainty/Distribution/MultivariatePolya.cxx]: fixed

File: lib/src/Uncertainty/Distribution/MultivariatePolya.cxx (hunk 25 above)
Comment (regislebrun) [reply lib/src/Uncertainty/Distribution/MultivariatePolya.cxx:845]: fixed

File: lib/src/Uncertainty/Distribution/MultivariateHypergeometric.cxx (hunk 26 above)
Comment (regislebrun) [reply lib/src/Uncertainty/Distribution/MultivariateHypergeometric.cxx:526]: Right: this recursion enumerates the whole marginal support, whose cardinal is bounded by prod_j (min(colors_[i_j], n_) + 1). It is pruned by the draw budget (xMax = min(colorJ, n_ - norm), and draws <= restTotal), so the work is proportional to the number of atoms the FiniteDiscreteDistribution must contain anyway; exponential only in the marginal size, which stays small in practice.

File: lib/src/Uncertainty/Distribution/MultivariateHypergeometric.cxx (hunk 27 above)
Comment (regislebrun) [reply lib/src/Uncertainty/Distribution/MultivariateHypergeometric.cxx]: fixed

File: lib/src/Uncertainty/Distribution/PoissonFormula.hxx (hunk 28 above)
Comment (regislebrun) [reply lib/src/Uncertainty/Distribution/PoissonFormula.hxx]: fixed: computeCDF is now templated on the phi functor

File: lib/src/Uncertainty/Distribution/PoissonFormula.hxx (hunk 29 above)
Comment (regislebrun) [reply lib/src/Uncertainty/Distribution/PoissonFormula.hxx]: fixed: computeProbability is now templated on the phi functor

File: lib/src/Uncertainty/Distribution/Multinomial.cxx
Comment (regislebrun) [reply lib/src/Uncertainty/Distribution/Multinomial.cxx:244]: As already handled: a zero p_i collapses the local generating function to 1 for windows starting at 0 and to 0 for windows above 0 (verified numerically on p=[0.5, 0.5, 0.0]); computePDF also guards the zero-remainder binomial via the sumP>0 test.

**Comment by regislebrun** [comment]
@coderabbit is your comment "The new distribution can return incorrect probabilities or marginals for supported API usage, so these issues should be resolved before merge." still true? Give actual examples of such issues

File: python/src/MultivariateHypergeometric_doc.i [hunk 30]
```diff
@@ -0,0 +1,183 @@
+%feature("docstring") OT::MultivariateHypergeometric
+R"RAW(Multivariate hypergeometric distribution.
+
+Randomizes drawing *n* balls without replacement from a finite population
+split into *d* categories of sizes :math:`(C_1, \ldots, C_d)`.
```
Comment (adutfoy) [inline python/src/MultivariateHypergeometric_doc.i]: Capital letters should be reserved to random variables. The sizes must be noted $c_i$ and not $C_i$!

File: python/src/MultivariateHypergeometric_doc.i [hunk 31]
```diff
@@ -0,0 +1,183 @@
...
+  computation;
+
+- *a_min*, the threshold below which the generating function of a truncated
+  binomial distribution is computed with its polynomial form rather than as a
+  correction of the untruncated generating function.
+
+Those two parameters can be tuned through the :class:`~openturns.ResourceMap`
+keys:
+
+- ``MultivariateHypergeometric-eta``, default :math:`1.0e-9`;
```
Comment (adutfoy) [inline python/src/MultivariateHypergeometric_doc.i]: Present the keys as done in the other classes: type, default values.

File: python/src/MultivariateHypergeometricFactory_doc.i [hunk 32]
```diff
@@ -0,0 +1,65 @@
+%feature("docstring") OT::MultivariateHypergeometricFactory
+R"RAW(Multivariate hypergeometric factory.
+
```
Comment (adutfoy) [inline python/src/MultivariateHypergeometricFactory_doc.i:3]: Idem: $C$ sould become $c$...

File: python/src/MultivariatePolya_doc.i [hunk 33]
```diff
@@ -0,0 +1,232 @@
...
+  computation;
+
+- *a_min*, the threshold below which the generating function of a truncated
+  negative binomial distribution is computed with its polynomial form rather
+  than as a correction of the untruncated generating function.
+
+Those two parameters can be tuned through the :class:`~openturns.ResourceMap`
+keys:
+
+- ``MultivariatePolya-eta``, default :math:`1.0e-9`;
```
Comment (adutfoy) [inline python/src/MultivariatePolya_doc.i]: Present keys as done in the other classes: type, default values

File: python/src/MultivariatePolya_doc.i [hunk 34]
```diff
@@ -0,0 +1,232 @@
+%feature("docstring") OT::MultivariatePolya
+R"RAW(Multivariate Polya distribution.
+
```
Comment (adutfoy) [inline python/src/MultivariatePolya_doc.i:3]: Is it possible to draw on the API a non trivial distribution?

File: python/src/MultivariatePolya_doc.i [hunk 35]
```diff
@@ -0,0 +1,232 @@
...
+        \Expect{X_i} & = & N \frac{\alpha_i}{q}, \quad i = 1, \ldots, d \\
+        \Var{X_i} & = & N \frac{\alpha_i}{q} \left(1 - \frac{\alpha_i}{q}\right) \frac{N + \theta}{1 + \theta}, \quad \theta = \frac{q}{c}, \quad i = 1, \ldots, d \\
+        \Cov{X_i, X_j} & = & - N \frac{\alpha_i \alpha_j}{q^2} \frac{N + \theta}{1 + \theta}, \quad i, j = 1, \ldots, d, i \neq j
+    \end{eqnarray*}
+
+where the overdispersion factor :math:`(N + \theta) / (1 + \theta)` is equal
+to 1 in the degenerate multinomial case :math:`c = 0`.
+
+The marginal distribution of a category is a beta-binomial
+:math:`\mathrm{X_i} \sim \mathrm{BetaBinomial}(N, \alpha_i / c, (q - \alpha_i) / c)`
```
Comment (adutfoy) [inline python/src/MultivariatePolya_doc.i]: can you give a reference for the BetaBinomial distribution? This one is not in the library: a wikipedia reference is needed!

File: python/src/MultivariatePolyaFactory_doc.i [hunk 36]
```diff
@@ -0,0 +1,76 @@
+%feature("docstring") OT::MultivariatePolyaFactory
+R"RAW(Multivariate Polya factory.
+
```
Comment (adutfoy) [inline python/src/MultivariatePolyaFactory_doc.i:3]: Is it possible to draw on the API a non trivial distribution?

File: python/src/MultinomialFactory_doc.i [hunk 37]
```diff
@@ -3,20 +3,24 @@ R"RAW(Multinomial factory.
 
 Notes
 -----
```
Comment (adutfoy) [inline python/src/MultinomialFactory_doc.i:5]: The mention of OpenTURNS should be replaced by 'the library'

**Review by adutfoy — CHANGES_REQUESTED** [review]

File: python/src/MultivariateHypergeometric_doc.i (hunk 30 above)
Comment (regislebrun) [reply python/src/MultivariateHypergeometric_doc.i]: Fixed: capital letters are now reserved for the random vector, the non-random quantities are written in lowercase (c_i for the sizes and c for the total population size), and the number of draws is written N.

File: python/src/MultivariateHypergeometric_doc.i (hunk 31 above)
Comment (regislebrun) [reply python/src/MultivariateHypergeometric_doc.i]: Fixed: the keys are now presented with their type (Scalar) and their default values, following the convention of the other classes.

File: python/src/MultivariateHypergeometricFactory_doc.i (hunk 32 above)
Comment (regislebrun) [reply python/src/MultivariateHypergeometricFactory_doc.i:3]: Fixed: C has been replaced by c in the factory documentation, and the number of draws is written N.

File: python/src/MultivariatePolya_doc.i (hunk 33 above)
Comment (regislebrun) [reply python/src/MultivariatePolya_doc.i]: Fixed: the keys are now presented with their type (Scalar) and their default values.

File: python/src/MultivariatePolya_doc.i (hunk 34 above)
Comment (regislebrun) [reply python/src/MultivariatePolya_doc.i:3]: Fixed: the Examples section now plots a non-trivial case, the PDF of a two-dimensional instance via drawPDF().

File: python/src/MultivariatePolya_doc.i (hunk 35 above)
Comment (regislebrun) [reply python/src/MultivariatePolya_doc.i]: Fixed: a Wikipedia reference for the beta-binomial distribution has been added.

File: python/src/MultivariatePolyaFactory_doc.i (hunk 36 above)
Comment (regislebrun) [reply python/src/MultivariatePolyaFactory_doc.i:3]: Fixed: the Examples section now plots the PDF of a two-dimensional estimate built from a sample.

File: python/src/MultinomialFactory_doc.i (hunk 37 above)
Comment (regislebrun) [reply python/src/MultinomialFactory_doc.i:5]: Fixed: the mention of OpenTURNS has been replaced by 'the library'.

File: python/src/MultivariatePolya_doc.i (hunk 34 above)
Comment (adutfoy) [reply python/src/MultivariatePolya_doc.i:3]: still no difference

File: python/src/MultivariatePolyaFactory_doc.i (hunk 36 above)
Comment (adutfoy) [reply python/src/MultivariatePolyaFactory_doc.i:3]: still no difference

**Comment by jschueller** [comment]
needs a proper rebase
